src/RevCPU.cc

```diff
@@ -2437,12 +2521,246 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
     rtn = false;
   }
 
-  if( rtn ){
+  if( rtn && CompletedThreads.size() ){
+    for( unsigned i=0; i<numCores; i++ ){
+      Procs[i]->PrintStatSummary();
+    }
     primaryComponentOKToEndSim();
-    output.verbose(CALL_INFO, 5, 0, "OK to end sim at cycle: %" PRIu64 "\n", currentCycle);
+    output.verbose(CALL_INFO, 5, 0, "OK to end sim at cycle: %" PRIu64 "\n", static_cast<uint64_t>(currentCycle));
+  } else {
+    rtn = false;
   }
 
   return rtn;
 }
 
+
+// Initializes a RevThread object.
+// - Moves it to the 'Threads' map 
+// - Adds it's ThreadID to the ThreadQueue to be scheduled
+void RevCPU::InitThread(std::shared_ptr<RevThread> ThreadToInit){
+
+  ThreadToInit->GetRegFile()->SetX(Procs[0]->GetRevFeature(), 3, Loader->GetSymbolAddr("__global_pointer$"));
+  ThreadToInit->GetRegFile()->SetX(Procs[0]->GetRevFeature(), 8, Loader->GetSymbolAddr("__global_pointer$"));
+
+  uint32_t TID = ThreadToInit->GetThreadID();
+  // Check if this ThreadID has already been assigned... if so... something has gone horribly wrong
+  // print out all threads
+  auto it = Threads.find(TID);
+  if( it != Threads.end() && it->second->GetState() != ThreadState::START ){
+    std::cout << *Threads.at(TID) << std::endl;
+    output.fatal(CALL_INFO, 99, "Error: ThreadID %d has already been assigned... this is a bug.\n", TID);
+  }
+  output.verbose(CALL_INFO, 4, 0, "Initializing Thread %d\n", TID);
+  output.verbose(CALL_INFO, 11, 0, "Thread Information: %s", ThreadToInit->print());
+  ThreadToInit->SetState(ThreadState::READY);
+  Threads.emplace(ThreadToInit->GetThreadID(), ThreadToInit);
+  ThreadQueue.emplace_back(TID);
+}
+
+void RevCPU::AssignThread(uint32_t ThreadID, uint32_t ProcID){
+  output.verbose(CALL_INFO, 4, 0, "Assigning Thread %" PRIu32 " to Processor %" PRIu32 "\n", ThreadID, ProcID);
+  auto Thread = Threads.at(ThreadID);
+  
+  // Point the regfile of this thread's LSQ to the Proc's LSQ
+  Thread->GetRegFile()->LSQueue = Procs[ProcID]->GetLSQueue();
+
```

Comment:
Why is the LSQueue assigned here? Wouldn’t this be part of the RegFile / Thread init?

Reply:
If it is assigned when the Thread is created in RevProc we don't know what Proc it will end up on (ie. What LSQueue it should be assigned to)